lib/IRGen/IRGenSIL.cpp

```diff
@@ -2053,38 +2053,87 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
     checkOk(status);
   };
 
-  // Handle inputs.
+  // Handle inputs and dynamic attributes.
   for (auto structuredArgument : opInfo.getStructuredArguments()) {
-    assert(structuredArgument.getArgumentNameWithSuffix().empty() &&
-           "cannot lower named arguments");
-    switch (structuredArgument.getKind()) {
-    case GraphOperationInfo::SAK_Single: {
-      LLVM_DEBUG(llvm::dbgs() << " Adding a single input.\n");
-      auto tensorHandle = structuredArgument.getSingleArgument();
-      addTensorhandleAsOpInput(tensorHandle);
+    auto argumentNameAndLowering =
+        structuredArgument.getArgumentNameAndLowering();
+    std::string argumentName = std::get<0>(argumentNameAndLowering);
+    auto argumentLowering = std::get<1>(argumentNameAndLowering);
+    LLVM_DEBUG(llvm::dbgs() << "IRGen graph_op argument " << argumentName
+               << "with lowering " << (unsigned)argumentLowering);
+    switch (argumentLowering) {
+    case GraphOperationInfo::ArgumentLowering::Out:
+      assert(0 && "Attributes cannot be output parameters");
+    case GraphOperationInfo::ArgumentLowering::UnknownShapeListAttribute:
+      assert(0 && "UnknownShapeListAttribute should have been eliminated "
+                  "by deabstraction");
+    case GraphOperationInfo::ArgumentLowering::TypeListAttribute:
+      assert(0 && "TypeListAttribute should have been eliminated by "
+                  "deabstraction");
+    case GraphOperationInfo::ArgumentLowering::TensorAttribute:
+      // See GraphOperationInfo.h for more information on why TensorAttribute
+      // cannot be dynamic.
+      assert(0 && "TensorAttributes cannot be dynamic");
+    case GraphOperationInfo::ArgumentLowering::Input: {
+      assert(argumentName.empty() && "inputs cannot have names");
```

Comment:
Would it make sense to refactor this into a separate function?

Reply:
Our style for these massive switch-cases in lowering seems to be to handle all the cases inline, and only factor out pieces of code that are shared between multiple cases (TFLowerGraph.cpp's visitGraphOperationInst, and also the "Handle const-evaluated attributes" later in this function). I'd like to keep it consistent for now.